src/mscorlib/src/System/String.cs

```diff
@@ -1208,6 +1272,42 @@ internal String[] SplitInternal(char[] separator, int count, StringSplitOptions
             }
             return foundCount;
         }
+
+        private struct SplitSeparatorArray
+        {
+            private readonly bool _single;
```

Comment:
Sigh, I need to take a break away from the screen...

Below should work without any extra checks:

```c#
    private struct SplitSeparatorArray
    {
        private readonly string _separator;
        private readonly string[] _separators;

        public SplitSeparatorArray(string separator)
        {
            _separator = separator ?? string.Empty;
            _separators = null;
        }

        public SplitSeparatorArray(string[] separators)
        {
            _separator = null;
            _separators = separators;
        }

        public bool IsNullOrEmpty
        {
            get { return _separator == null && (_separators == null || _separators.Length == 0); }
        }

        public int Length
        {
            get { return _separator != null ? 1 : _separators.Length; }
        }

        public string this[int index]
        {
            get { return _separator != null ? _separator : _separators[index]; }
        }
    }
```